### PR TITLE
[Copilot] Azure OpenAI Tools support added to our Copilot Toolkit

### PR DESCRIPTION
**Pull request by qasimikram**
#### Summary 

This pull request introduces Azure OpenAI Tools to our Copilot Toolkit. Currently, this functionality is open for application by Microsoft only. We will open this for community when next GA version of Azure Open AI is available which supports tools.

#### Work Item(s) 

[AB#487896](https://dynamicssmb2.visualstudio.com/Dynamics%20SMB/_workitems/edit/487896)

### CODE REVIEW

File: src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIToolsImpl.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            Error(ErrorMessage);
+        end;
+        exit(true);
+    end;
+
+    [NonDebuggable]
+    local procedure CapabilityBlocked(CallerModuleInfo: ModuleInfo)
+    var
+        CurrentModuleInfo: ModuleInfo;
+        CapabilityBlockedErr: Label 'Tools capability is currently blocked for non Microsoft features.';
```
Comment (Groenbech96) [inline src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIToolsImpl.Codeunit.al]: label should me under global vars

**Review by Groenbech96 — DISMISSED** [review]
Fix minor, but LGTM

File: src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIChatMessages.Codeunit.al [hunk 2]
```diff
@@ -176,4 +177,103 @@ codeunit 7763 "AOAI Chat Messages"
...
+    begin
+        AOAIToolsImpl.ModifyTool(Id, NewTool);
+    end;
+
+    /// <summary>
+    /// Deletes a Tool from the list of Tool.
+    /// </summary>
+    /// <param name="Id">Id of the Tool.</param>
+    /// <error>Message id does not exist.</error>
+    [NonDebuggable]
```
Comment (darjoo) [inline src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIChatMessages.Codeunit.al]: Remove debuggable

File: src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIChatMessages.Codeunit.al [hunk 3]
```diff
@@ -176,4 +177,103 @@ codeunit 7763 "AOAI Chat Messages"
...
+    procedure GetTools(): List of [JsonObject]
+    begin
+        exit(AOAIToolsImpl.GetTools());
+    end;
+
+    /// <summary>
+    /// Checks if at least one Tools exists in the list.
+    /// </summary>
+    /// <returns>True if Tools exists, false otherwise.</returns>
+    [NonDebuggable]
```
Comment (darjoo) [inline src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIChatMessages.Codeunit.al]: Remove debuggable

File: src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIChatMessages.Codeunit.al [hunk 4]
```diff
@@ -176,4 +177,103 @@ codeunit 7763 "AOAI Chat Messages"
...
+    procedure ToolExists(): Boolean
```
Comment (darjoo) [inline src/System Application/App/AI/src/Azure OpenAI/Chat Completion/AOAIChatMessages.Codeunit.al]: Rename to `ToolsExists`. Using `ToolExists` sounds like I can check if a specific tool exists.

**Review by darjoo — COMMENTED** [review]
Additionally, no test for any of it? :(

**Comment by qasimikram** [comment]
> Additionally, no test for any of it? :(

Tests added.

**Comment by qasimikram** [comment]
> Fix minor, but LGTM

Done.

File: src/System Application/Test/AI/src/AzureOpenAIToolsTest.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    var
+        TestTool: Text;
+        ToolJsonObject: JsonObject;
+    begin
+        TestTool := '{"type": "function", "function": {"name": "test_function_1", "parameters": {"type": "object", "properties": {"message": {"type": "string", "description": "The input from user."}}}}}';
+        ToolJsonObject.ReadFrom(TestTool);
+        exit(ToolJsonObject);
+    end;
+
+    local procedure GetTestFunction2Tool(): JsonObject
```
Comment (darjoo) [inline src/System Application/Test/AI/src/AzureOpenAIToolsTest.Codeunit.al:131]: Can you add 1 more test that tests the `PrepareTools` functionality? 
That can be accessed by creating another function like the current one in the test library.

**Review by darjoo — APPROVED** [review]

File: src/System Application/Test/AI/src/AzureOpenAIToolsTest.Codeunit.al (hunk 5 above)
Comment (qasimikram) [reply src/System Application/Test/AI/src/AzureOpenAIToolsTest.Codeunit.al:131]: Test added.

**Review by darjoo — APPROVED** [review]

**Review by Groenbech96 — APPROVED** [review]

**Review by JesperSchulz — APPROVED** [review]

**Review by darjoo — APPROVED** [review]

**Pull request merged**